File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/security/extractor/JwtHeaderTokenExtractor.java
package com.ntnu.gidd.security.extractor;


import com.ntnu.gidd.security.config.JWTConfig;
import lombok.AllArgsConstructor;
import org.springframework.security.authentication.AuthenticationServiceException;
import org.springframework.stereotype.Component;


/**
 * Extracts token from a header with the attribute:
 * Authorization: Bearer <token>
 */
@Component
@AllArgsConstructor
public class JwtHeaderTokenExtractor implements TokenExtractor {

    private JWTConfig jwtConfig;

    /**
     * Validate the header and extract the jwt token.
     * @param header the header containing the jwt token
     * @return the jwt token substring
     */
    @Override
    public String extract(String header) {
        if (header == null || header.isBlank())
            throw new AuthenticationServiceException("Authorization header cannot be blank.");

        if (header.length() < jwtConfig.getHeader().length())
            throw new AuthenticationServiceException("Invalid authorization header size.");

        return header.substring(jwtConfig.getPrefix().length());
    }
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/extractor/JwtHeaderTokenExtractor.kt
package ntnu.idatt2105.security.extractor

import ntnu.idatt2105.security.config.JWTConfig
import org.springframework.security.authentication.AuthenticationServiceException
import org.springframework.stereotype.Component

@Component
class JwtHeaderTokenExtractor(val jwtConfig: JWTConfig) : TokenExtractor {
    override fun extract(payload: String): String {
        if (payload.isBlank()) throw AuthenticationServiceException("Authorization header cannot be blank.")
        if (payload.length < jwtConfig.header.length) throw AuthenticationServiceException("Invalid authorization header size.")
        return payload.substring(jwtConfig.prefix.length)
    }
}
